### PR TITLE
Implement time_point and time_duration classes:

### PR DESCRIPTION
**Pull request by BevapDin**
To represent some duration of time / some point in time.

Adds various operators to make them behave naturally.

Adds user defined literals to create `time_duration` instances from literals (e.g. `10_turns`, `5_hours`).

Changes `item::bday` to `time_point` and associated code to use the new types.

### CODE REVIEW

File: src/rng.h [hunk 1]
```diff
@@ -7,6 +7,9 @@
 #include <functional>
 
 long rng( long val1, long val2 );
+class time_duration;
+// implemented in calendar.cpp
+time_duration rng( time_duration val1, time_duration val2 );
```
Comment (kevingranade) [inline src/rng.h]: Just move the declaration to calendar.h too, no need to rng to know about this type.

**Comment by kevingranade** [comment]
What's the payoff here?  This is a big chunk of boilerplate that we're going to haul around indefinitely, and a decent chunk of code churn just to extract a handful of constants to a wrapper class.  Plus the post-modification code isn't significantly more natural than before (except the few places where it was using anumber in turns already instead of using the MINUTES() macro.

(Note I fully expect that you have a goal in mind, I just want to know what it is :)

File: src/calendar.h [hunk 2]
```diff
@@ -342,4 +342,286 @@ class calendar
         static season_type initial_season;
 };
 
+class time_duration;
```
Comment (kevingranade) [inline src/calendar.h]: I think this is extensive enough to merit its own header, even if calendar.h just ends up including it.
It looks like quite a few files in this PR are adding a dependency on calendar.h when they only need time.h

**Comment by BevapDin** [comment]
> What's the payoff here?

Proper type safety. A time duration is not the same thing as a time point, neither are the same as a plain integer. It's basically the same as for `units::volume` and `units::mass`.

> Plus the post-modification code isn't significantly more natural than before (except the few places where it was using anumber in turns already instead of using the MINUTES() macro.

Even the `MINUTES` function is not good enough. It still returns a plain integer that nobody knows what it represents. And these functions are not consistent: `MOVES` returns the number of moves and expects the number of turns as input. `SECONDS` expects the number of seconds and returns the number of turns.

The wrapper class abstracts that completely away: `time_duration::from_hours( 10 )` (or `10_hours`) yields an object that represents "10 hours" and can be compared to `600_minutes`, without knowing the internals (which could even store the duration in moves or seconds).

----

Another thing this makes clear is the distinction between duration and point (age and birthday for items): Look at the various functions in "activity_handlers.cpp" that deal with butchering - they have a parameter `int age`, but it's actually used as item *birthday* (not *age*). This inconsistency if far more obvious with proper types (`time_point age` sounds wrong).

Calls to functions like `add_effect` are much easier to read:
```C++
u.add_effect( "poisoned", 20 ); // what does 20 mean?
u.add_effect( "poisoned", 20_turns ); // much clearer.
```

----

[This](https://github.com/BevapDin/Cataclysm-DDA/tree/time) is a preview for the final code. Very rough and not complete and does not even compile.

----

> I think this is extensive enough to merit its own header, even if calendar.h just ends up including it.
> It looks like quite a few files in this PR are adding a dependency on calendar.h when they only need time.h

I plan to move all functionality of `calendar` into the two new classes or into a namespace `calendar` (for example the season length function). Those functions and the static data from the namespace constitute the calendar ("A calendar is a system of organizing days for social, religious, commercial or administrative purposes." - Wikipedia). The use of "calendar" to refer to a specific point in time is a bit odd.

**Comment by kevingranade** [comment]
> I plan to move all functionality of calendar into the two new classes or into a namespace calendar (for example the season length function). Those functions and the static data from the namespace constitute the calendar ("A calendar is a system of organizing days for social, religious, commercial or administrative purposes." - Wikipedia). The use of "calendar" to refer to a specific point in time is a bit odd.

What I mean is it looks like many times as many compilation units need to pass around and compare time points or time spans compared to files that need to call into the existing calendar methods.  No need to burden these callers with all the calendar module details if they just need to use the time data types.

Analogy, this is like requiring map.h in order to use a tripoint, a lot of code can utilize a tripoint without referencing the map object.

**Comment by BevapDin** [comment]
I moved the definition of `time_point` and `time_duration` into "time.h" / "time.cpp" - but I don't like it. "calendar.h" will end up being nearly empty. Those classes form the calendar system, so it makes sense (in my opinion) to have them there.

Anyway, this now causes the compilation to fail. Apparently the compiler tries to include the new "time.h" from within the system header "ctime". And it later complains that several things like `ctime` are not declared (they used to be declared via the system header "time.h"). What now?

**Comment by kevingranade** [comment]
I guess it has a name collision with a system library, needs a
non-colliding name, time_units.h?

If you think the inclusion overhead is insignificant and/or keeping the
calendar code together is more important, feel free to reset to the
previous commit or just say so and I'll pull that one,  I just wanted to
make sure my concern was clear, I wasn't trying to insist on it.

**Pull request merged**